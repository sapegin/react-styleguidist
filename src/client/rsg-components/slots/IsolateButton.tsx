import React from 'react';
import PropTypes from 'prop-types';
import { MdFullscreen, MdFullscreenExit } from 'react-icons/md';
import ToolbarButton from 'rsg-components/ToolbarButton';

export interface IsolateButtonProps {
	name: string;
	href: string;
	isolated?: boolean;
	exampleIndex?: number | string;
}

const IsolateButton = ({ name, href, isolated, exampleIndex }: IsolateButtonProps) => {
	// TODO: Rewrite Cypress tests to use Cypress Testing Library and remove this ID
	const testID = exampleIndex ? `${name}-${exampleIndex}-isolate-button` : `${name}-isolate-button`;

	return isolated ? (
		<ToolbarButton href={href} title="Show all components" testId={testID}>
			<MdFullscreenExit />
		</ToolbarButton>
	) : (
		<ToolbarButton href={href} title="Open isolated" testId={testID}>
			<MdFullscreen />
		</ToolbarButton>
	);
};

IsolateButton.propTypes = {
	name: PropTypes.string.isRequired,
	href: PropTypes.string.isRequired,
	isolated: PropTypes.bool,
	exampleIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
};

export default IsolateButton;
