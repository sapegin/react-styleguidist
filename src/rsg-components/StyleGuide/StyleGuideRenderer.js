import React, { PropTypes } from 'react';
import Logo from 'rsg-components/Logo';
import Markdown from 'rsg-components/Markdown';
import Styled from 'rsg-components/Styled';
import cx from 'classnames';

const styles = ({ font, base, light, border, baseBackground, codeBackground }) => ({
	root: {
		isolate: false,
		color: base,
		backgroundColor: baseBackground,
	},
	hasSidebar: {
		isolate: false,
		paddingLeft: 200,
	},
	content: {
		isolate: false,
		maxWidth: 1000,
		padding: [[15, 30]],
		margin: [[0, 'auto']],
	},
	sidebar: {
		backgroundColor: codeBackground,
		borderRight: [[1, border, 'solid']],
		position: 'fixed',
		top: 0,
		left: 0,
		bottom: 0,
		width: 200,
		overflow: 'scroll',
	},
	components: {
		isolate: false,
		overflow: 'auto',  // To prevent the pane from growing out of the screen
	},
	logo: {
		padding: 15,
		borderBottom: [[1, border, 'solid']],
	},
	footer: {
		color: light,
		fontFamily: font,
		fontSize: 12,
	},
});

export const StyleGuideRenderer = ({ classes, title, homepageUrl, components, toc, sidebar }) => (
	<div className={cx(classes.root, sidebar && classes.hasSidebar)}>
		<main className={classes.content}>
			<div className={classes.components}>
				{components}
				<footer className={classes.footer}>
					<Markdown text={`Generated with [React Styleguidist](${homepageUrl})`} />
				</footer>
			</div>
		</main>
		{sidebar &&
			<div className={classes.sidebar}>
				<div className={classes.logo}>
					<Logo>{title}</Logo>
				</div>
				{toc}
			</div>
		}
	</div>
);

StyleGuideRenderer.propTypes = {
	classes: PropTypes.object.isRequired,
	title: PropTypes.string.isRequired,
	homepageUrl: PropTypes.string.isRequired,
	components: PropTypes.object.isRequired,
	toc: PropTypes.node.isRequired,
	sidebar: PropTypes.bool,
};

export default Styled(styles)(StyleGuideRenderer);
