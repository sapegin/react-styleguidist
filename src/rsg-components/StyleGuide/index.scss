.main {
  position: relative;
  max-width: 200000px;
  padding: 0px 0px 46px 32px;
  margin: 0 auto;
  display: block;
}

.footer {
  width: 100%;
  position: absolute;
  bottom: 0;
  left: 0;
  background: whitesmoke;
  padding: 12px 32px;
  border: 1px solid #e8e8e8;
  border-width: 1px 0px 0px 0px;

  p {
	margin: 0;
  }
}
