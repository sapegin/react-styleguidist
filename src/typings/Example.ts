export interface Example {
	type: 'code' | 'text';
	lang: string;
	content: string;
	settings: { [key: string]: string };
}
