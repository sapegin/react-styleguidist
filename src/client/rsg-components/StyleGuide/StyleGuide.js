import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableOfContents from 'rsg-components/TableOfContents';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import Welcome from 'rsg-components/Welcome';
import Error from 'rsg-components/Error';
import NotFound from 'rsg-components/NotFound';
import Context from 'rsg-components/Context';
import { HOMEPAGE } from '../../../scripts/consts';
import { DisplayModes } from '../../consts';

/**
 * This function will return true, if the sidebar should be visible and false otherwise.
 *
 * These sorted conditions (highest precedence first) define the visibility
 * state of the sidebar.
 *
 * - Sidebar is hidden for isolated example views
 * - Sidebar is always visible when pagePerSection
 * - Sidebar is hidden when showSidebar is set to false
 * - Sidebar is visible when showSidebar is set to true for non-isolated views
 *
 * @param {boolean} displayMode
 * @param {boolean} showSidebar
 * @param {boolean} pagePerSection
 * @returns {boolean}
 */
function hasSidebar(displayMode, showSidebar) {
	return displayMode === DisplayModes.notFound || (showSidebar && displayMode === DisplayModes.all);
}

export default class StyleGuide extends Component {
	static propTypes = {
		codeRevision: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		patterns: PropTypes.array,
		displayMode: PropTypes.string,
		allSections: PropTypes.array.isRequired,
		pagePerSection: PropTypes.bool,
	};
	static defaultProps = {
		displayMode: DisplayModes.all,
	};

	state = {
		error: false,
		info: null,
	};

	componentDidCatch(error, info) {
		this.setState({
			error,
			info,
		});
	}

	render() {
		const { error, info } = this.state;
		const {
			config,
			sections,
			welcomeScreen,
			patterns,
			displayMode,
			allSections,
			pagePerSection,
			codeRevision,
			slots,
		} = this.props;

		if (error) {
			return <Error error={error} info={info} />;
		}

		if (welcomeScreen) {
			return <Welcome patterns={patterns} />;
		}

		return (
			<Context.Provider
				value={{
					codeRevision,
					config,
					slots,
					displayMode,
				}}
			>
				<StyleGuideRenderer
					title={config.title}
					version={config.version}
					homepageUrl={HOMEPAGE}
					toc={<TableOfContents sections={allSections} useRouterLinks={pagePerSection} />}
					hasSidebar={hasSidebar(displayMode, config.showSidebar)}
				>
					{sections.length ? <Sections sections={sections} depth={1} /> : <NotFound />}
				</StyleGuideRenderer>
			</Context.Provider>
		);
	}
}
