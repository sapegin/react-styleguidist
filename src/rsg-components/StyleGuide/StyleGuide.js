import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableOfContents from 'rsg-components/TableOfContents';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import Welcome from 'rsg-components/Welcome';
import { HOMEPAGE } from '../../../scripts/consts';
import { filterSectionsByName } from '../../utils/utils';

export default class StyleGuide extends Component {
	static propTypes = {
		codeKey: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		patterns: PropTypes.array,
		isolatedComponent: PropTypes.bool,
		isolatedExample: PropTypes.bool,
		isolatedSection: PropTypes.bool,
	};

	static childContextTypes = {
		codeKey: PropTypes.number.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		isolatedComponent: PropTypes.bool,
		isolatedExample: PropTypes.bool,
		isolatedSection: PropTypes.bool,
	};

	static defaultProps = {
		isolatedComponent: false,
	};

	state = {
		searchTerm: '',
	};

	getChildContext() {
		return {
			codeKey: this.props.codeKey,
			config: this.props.config,
			slots: this.props.slots,
			isolatedComponent: this.props.isolatedComponent,
			isolatedExample: this.props.isolatedExample,
			isolatedSection: this.props.isolatedSection,
		};
	}

	render() {
		const { config, sections, welcomeScreen, patterns, isolatedComponent } = this.props;

		const { searchTerm } = this.state;
		const filteredSections = (0, filterSectionsByName)(sections, searchTerm);

		if (welcomeScreen) {
			return <Welcome patterns={patterns} />;
		}

		return (
			<StyleGuideRenderer
				title={config.title}
				homepageUrl={HOMEPAGE}
				hasSidebar={config.showSidebar && !isolatedComponent}
				toc={
					<TableOfContents
						sections={filteredSections}
						searchTerm={searchTerm}
						onSearchTermChange={searchTerm => this.setState({ searchTerm })}
					/>
				}
			>
				<Sections sections={filteredSections} />
			</StyleGuideRenderer>
		);
	}
}
