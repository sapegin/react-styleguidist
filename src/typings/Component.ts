import { Example } from './Example';

export interface Component {
	name?: string;
	visibleName?: string;
	props: {
		displayName: string;
		visibleName?: string;
		examples?: Example[];
		example?: Example[];
	};
	module?: number;
}
