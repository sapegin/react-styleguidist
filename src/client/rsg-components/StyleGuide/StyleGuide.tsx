import React, { Component } from 'react';
import PropTypes from 'prop-types';
import TableOfContents from 'rsg-components/TableOfContents';
import StyleGuideRenderer from 'rsg-components/StyleGuide/StyleGuideRenderer';
import Sections from 'rsg-components/Sections';
import Welcome from 'rsg-components/Welcome';
import Error from 'rsg-components/Error';
import NotFound from 'rsg-components/NotFound';
import Context from 'rsg-components/Context';
import MdxProvider from 'rsg-components/mdx/MdxProvider';
import { HOMEPAGE } from '../../../scripts/consts';
import * as Rsg from '../../../typings';

/**
 * This function will return true, if the sidebar should be visible and false otherwise.
 *
 * These sorted conditions (highest precedence first) define the visibility
 * state of the sidebar.
 *
 * - Sidebar is always hidden in isolated mode
 * - Sidebar is visible when pagePerSection
 * - Sidebar is hidden when showSidebar is set to false
 * - Sidebar is visible when showSidebar is set to true for non-isolated views
 */
function hasSidebar(sections: Rsg.Section[], isolated: boolean, showSidebar: boolean): boolean {
	return sections.length > 0 && !isolated && showSidebar;
}

export interface StyleGuideProps {
	codeRevision: number;
	cssRevision: string;
	config: Rsg.ProcessedStyleguidistConfig;
	slots: any;
	sections: Rsg.Section[];
	welcomeScreen?: boolean;
	patterns?: string[];
	isolated: boolean;
	exampleIndex?: number | string;
	allSections?: Rsg.Section[];
	pagePerSection?: boolean;
}

interface StyleGuideState {
	error: Error | boolean;
	info: React.ErrorInfo | null;
}

export default class StyleGuide extends Component<StyleGuideProps, StyleGuideState> {
	public static propTypes = {
		codeRevision: PropTypes.number.isRequired,
		cssRevision: PropTypes.string.isRequired,
		config: PropTypes.object.isRequired,
		slots: PropTypes.object.isRequired,
		sections: PropTypes.array.isRequired,
		welcomeScreen: PropTypes.bool,
		patterns: PropTypes.array,
		isolated: PropTypes.bool.isRequired,
		exampleIndex: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
		allSections: PropTypes.array.isRequired,
		pagePerSection: PropTypes.bool,
	};

	public state = {
		error: false,
		info: null,
	};

	public componentDidCatch(error: Error, info: React.ErrorInfo) {
		this.setState({
			error,
			info,
		});
	}

	public render() {
		const { error, info }: StyleGuideState = this.state;
		const {
			config,
			sections,
			welcomeScreen,
			patterns,
			isolated,
			exampleIndex,
			allSections,
			pagePerSection,
			codeRevision,
			cssRevision,
			slots,
		} = this.props;

		if (error && info) {
			return <Error error={error} info={info} />;
		}

		if (welcomeScreen && patterns) {
			return <Welcome patterns={patterns} />;
		}

		return (
			<Context.Provider
				value={{
					codeRevision,
					config,
					slots,
					isolated,
					exampleIndex,
					cssRevision,
				}}
			>
				<MdxProvider>
					<StyleGuideRenderer
						key={cssRevision}
						title={config.title}
						version={config.version}
						homepageUrl={HOMEPAGE}
						toc={
							allSections ? (
								<TableOfContents
									sections={allSections}
									pagePerSection={pagePerSection}
									tocMode={config.tocMode}
								/>
							) : null
						}
						hasSidebar={hasSidebar(sections, isolated, config.showSidebar)}
					>
						{sections.length ? <Sections sections={sections} depth={1} /> : <NotFound />}
					</StyleGuideRenderer>
				</MdxProvider>
			</Context.Provider>
		);
	}
}
