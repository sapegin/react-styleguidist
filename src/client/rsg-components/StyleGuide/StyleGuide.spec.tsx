import React from 'react';
import { render, within } from '../../test';
import StyleGuide, { StyleGuideProps } from './StyleGuide';
import slots from '../slots';
import * as Rsg from '../../../typings';

const sections: Rsg.Section[] = [
	{
		exampleMode: 'collapse',
		usageMode: 'collapse',
		name: 'Section',
		visibleName: 'Section',
		slug: 'section',
		hashPath: ['Section'],
		sections: [],
		components: [
			{
				name: 'Foo',
				visibleName: 'Foo',
				hashPath: ['Foo'],
				slug: 'foo',
				pathLine: 'components/foo.js',
				filepath: 'components/foo.js',
				props: {
					displayName: 'Foo',
				},
				metadata: {},
				hasExamples: false,
			},
			{
				name: 'Bar',
				visibleName: 'Bar',
				hashPath: ['Bar'],
				slug: 'bar',
				pathLine: 'components/bar.js',
				filepath: 'components/bar.js',
				props: {
					displayName: 'Bar',
				},
				metadata: {},
				hasExamples: false,
			},
		],
	},
];

const config = {
	title: 'HelloStyleGuide',
	version: '1.0.0',
	showSidebar: true,
	pagePerSection: false,
} as Rsg.ProcessedStyleguidistConfig;

const defaultProps: StyleGuideProps = {
	codeRevision: 1,
	cssRevision: '1',
	config,
	pagePerSection: false,
	sections: [],
	allSections: [],
	slots: slots(),
	isolated: false,
	patterns: ['components/**.js'],
};

test('should render components', () => {
	const { getByText } = render(
		<StyleGuide {...defaultProps} sections={sections} allSections={sections} />
	);
	expect(getByText('components/foo.js')).toBeInTheDocument();
	expect(getByText('components/bar.js')).toBeInTheDocument();
});

test('should render welcome screen', () => {
	const { getByText } = render(<StyleGuide {...defaultProps} welcomeScreen />);
	expect(getByText('Welcome to React Styleguidist!')).toBeInTheDocument();
});

test('should render a sidebar if showSidebar is not set', () => {
	const { getByTestId } = render(
		<StyleGuide {...defaultProps} sections={sections} allSections={sections} />
	);
	const sidebar = within(getByTestId('sidebar'));
	const links = sidebar.getAllByRole('link');
	expect(links.map((node: any) => node.href)).toEqual([
		'http://localhost/#foo',
		'http://localhost/#bar',
	]);
	expect(links.map((node) => node.textContent)).toEqual(['Foo', 'Bar']);
});

test('should not render a sidebar if showSidebar is false', () => {
	const { queryByTestId } = render(
		<StyleGuide
			{...defaultProps}
			config={{
				...config,
				showSidebar: false,
			}}
			sections={sections}
			allSections={sections}
		/>
	);
	expect(queryByTestId('sidebar')).not.toBeInTheDocument();
});

test('should not render a sidebar in isolated mode', () => {
	const { queryByTestId } = render(
		<StyleGuide {...defaultProps} sections={sections} allSections={sections} isolated />
	);
	expect(queryByTestId('sidebar')).not.toBeInTheDocument();
});

test('should render a sidebar if pagePerSection is true', () => {
	const { getByTestId } = render(
		<StyleGuide {...defaultProps} sections={sections} allSections={sections} pagePerSection />
	);
	expect(getByTestId('sidebar')).toBeInTheDocument();
});

describe('error handling', () => {
	const console$error = console.error;
	beforeAll(() => {
		console.error = jest.fn();
	});
	afterAll(() => {
		console.error = console$error;
	});
	test('should render an error when componentDidCatch() is triggered', () => {
		const { getByText } = render(
			<StyleGuide {...defaultProps} patterns={null as any} welcomeScreen />
		);
		expect(getByText(/Page not found/i)).toBeInTheDocument();
	});
});
