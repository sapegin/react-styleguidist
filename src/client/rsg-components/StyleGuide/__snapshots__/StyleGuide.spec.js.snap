// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer should render logo, version, table of contents, ribbon and passed children 1`] = `
<div
  className=""
>
  <main>
    <h1>
      Content
    </h1>
    <footer>
      <Markdown
        text="Created with [React Styleguidist](http://react-styleguidist.js.org/)"
      />
    </footer>
  </main>
  <div
    data-testid="sidebar"
  >
    <header>
      <Styled(Logo)>
        Hello
      </Styled(Logo)>
      <Styled(Version)>
        1.0.0
      </Styled(Version)>
    </header>
    <TableOfContents
      sections={
        Array [
          Object {
            "components": Array [
              Object {
                "filepath": "components/foo.js",
                "name": "Foo",
                "pathLine": "components/foo.js",
                "props": Object {
                  "description": "Foo foo",
                },
              },
              Object {
                "filepath": "components/bar.js",
                "name": "Bar",
                "pathLine": "components/bar.js",
                "props": Object {
                  "description": "Bar bar",
                },
              },
            ],
          },
        ]
      }
    />
  </div>
  <Ribbon />
</div>
`;

exports[`should render an error when componentDidCatch() is triggered 1`] = `
<Styled(Error)
  error={
    Object {
      "toString": [Function],
    }
  }
  info={
    Object {
      "componentStack": Object {
        "toString": [Function],
      },
    }
  }
/>
`;

exports[`should render components list 1`] = `
<Styled(StyleGuide)
  hasSidebar={true}
  homepageUrl="https://react-styleguidist.js.org/"
  title="Hello"
  toc={
    <TableOfContents
      sections={
        Array [
          Object {
            "components": Array [
              Object {
                "filepath": "components/foo.js",
                "name": "Foo",
                "pathLine": "components/foo.js",
                "props": Object {
                  "description": "Foo foo",
                },
              },
              Object {
                "filepath": "components/bar.js",
                "name": "Bar",
                "pathLine": "components/bar.js",
                "props": Object {
                  "description": "Bar bar",
                },
              },
            ],
          },
        ]
      }
      useRouterLinks={false}
    />
  }
  version="1.0.0"
>
  <Sections
    depth={1}
    sections={
      Array [
        Object {
          "components": Array [
            Object {
              "filepath": "components/foo.js",
              "name": "Foo",
              "pathLine": "components/foo.js",
              "props": Object {
                "description": "Foo foo",
              },
            },
            Object {
              "filepath": "components/bar.js",
              "name": "Bar",
              "pathLine": "components/bar.js",
              "props": Object {
                "description": "Bar bar",
              },
            },
          ],
        },
      ]
    }
  />
</Styled(StyleGuide)>
`;

exports[`should render welcome screen 1`] = `<Styled(Welcome) />`;
